server/pairing/challenge_management.go

```diff
@@ -0,0 +1,247 @@
+		cg.logger.Error("regenerateNewChallenge: err = s.Save(r, w)", zap.Error(err))
+		return nil, &ChallengeError{"error", http.StatusInternalServerError}
+	}
+
+	return challenge, nil
+}
+
+func (cg *ChallengeGiver) block(s *sessions.Session, w http.ResponseWriter, r *http.Request) error {
+	s.Values[sessionBlocked] = true
+	err := s.Save(r, w)
+	if err != nil {
+		cg.logger.Error("block: err = s.Save(r, w)", zap.Error(err))
+		return &ChallengeError{"error", http.StatusInternalServerError}
+	}
+
+	return &ChallengeError{"forbidden", http.StatusForbidden}
+}
+
+func (cg *ChallengeGiver) checkChallengeResponse(w http.ResponseWriter, r *http.Request) error {
+	err := cg.validateClientIP(r)
+	if err != nil {
+		return err
+	}
+
+	s, err := cg.getSession(r)
+	if err != nil {
+		return err
+	}
+
+	blocked, ok := s.Values[sessionBlocked].(bool)
+	if ok && blocked {
+		return &ChallengeError{"forbidden", http.StatusForbidden}
+	}
+
+	// If the request header doesn't include a challenge don't punish the client, just throw a 403
+	clientChallengeResp := r.Header.Get(sessionChallenge)
+	if clientChallengeResp == "" {
+		return &ChallengeError{"forbidden", http.StatusForbidden}
+	}
+
+	dcr, err := cg.encryptor.decryptPlain(base58.Decode(clientChallengeResp))
+	if err != nil {
+		cg.logger.Error("checkChallengeResponse: cg.encryptor.decryptPlain(base58.Decode(clientChallengeResp))", zap.Error(err), zap.String("clientChallengeResp", clientChallengeResp))
+		return &ChallengeError{"error", http.StatusInternalServerError}
+	}
+
+	// If the challenge is not in the session store don't punish the client, just throw a 403
+	challenge, ok := s.Values[sessionChallenge].([]byte)
+	if !ok {
+		return &ChallengeError{"forbidden", http.StatusForbidden}
+	}
+
+	// Only if we have both a challenge in the session store and in the request header
+	// do we entertain blocking the client. Because then we know someone is trying to be sneaky.
+	if !bytes.Equal(dcr, challenge) {
+		return cg.block(s, w, r)
+	}
+
+	// If every is ok, generate a new challenge for the next req
+	_, err = cg.generateNewChallenge(s, w, r)
+	return err
+}
+
+func (cg *ChallengeGiver) getChallenge(w http.ResponseWriter, r *http.Request) ([]byte, error) {
+	err := cg.validateClientIP(r)
+	if err != nil {
+		return nil, err
+	}
+
+	s, err := cg.getSession(r)
+	if err != nil {
+		return nil, err
+	}
+
+	challenge, ok := s.Values[sessionChallenge].([]byte)
+	if !ok {
+		challenge, err = cg.generateNewChallenge(s, w, r)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return challenge, nil
+}
+
+// ChallengeTaker is responsible for storing and performing server challenges
+type ChallengeTaker struct {
+	encryptor       *PayloadEncryptor
+	serverChallenge []byte
+}
+
+func NewChallengeTaker(e *PayloadEncryptor) *ChallengeTaker {
+	return &ChallengeTaker{
+		encryptor: e.Renew(),
+	}
+}
+
+func (ct *ChallengeTaker) SetChallenge(resp *http.Response) error {
+	challenge, err := ioutil.ReadAll(resp.Body)
```

Comment:
Perhaps I need to change the name of this function `SetChallenge`, or better explain the context. `SetChallenge` takes a `http.Response` and sets the internal field `ct.serverChallenge`.

The function doesn't set the `http.Request` challenge, `ChallengeTaker` sets its own internal value from a `http.Response`. I can add a comment on this function to better explain it. Does that make sense @qfrank?

Reply:
sorry, maybe my previous comment is not clear. i mean we can get challenge by parsing from http header e.g. `resp.Header.Get("Set-Cookie")`, we don't need read http response body since challenge already being included in http header ? @Samyoul

Reply:
there is already a place used for challenge to return, now http body also contains it, this(> 1 places) makes me confused

Reply:
Hmm @qfrank , I'm confused now because we don't use `Set-Cookie` anywhere in the code. Well at least I can't see it. We've always took the challenge from the response body, because the value in the cookie is encrypted with a random key.

---

I did a little bit of research on this point just to be certain, I added some code to catch the cookie value and the challenge:

```go
func (c *BaseClient) getChallenge() error {
	c.baseAddress.Path = pairingChallenge
	resp, err := c.Get(c.baseAddress.String())
	if err != nil {
		return err
	}

	if resp.StatusCode != http.StatusOK {
		return fmt.Errorf("[client] status not ok when getting challenge, received '%s'", resp.Status)
	}

	// START New code
	cs := resp.Cookies()
	for _, c := range cs {
		fmt.Printf("  %s=%s\n", c.Name, c.Value)
		decodedBytes, err := base64.StdEncoding.DecodeString(c.Value)
		if err != nil {
			fmt.Println("Error decoding string:", err)
		}
		spew.Dump(decodedBytes)
	}

	spew.Dump(resp.Header.Get("Set-Cookie"))
	// END New code

	return c.challengeTaker.SetChallenge(resp)
}
```

```go
func (ct *ChallengeTaker) SetChallenge(resp *http.Response) error {
	challenge, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		return err
	}
	spew.Dump("SetChallenge", challenge)
	ct.serverChallenge = challenge
	return nil
}
```

The output was:

```text
challenge=MTY3OTU4NjIyN3x1cHJrVDVRUmlwMHEzbFdTc3YtR24xYUoxcmpWbWJPVVkxekVHVG9IR0pVX3B0Q3lDUzdwWklsRVZMbWlyaGZkckpRSUVuRGRPdUtKNW1nRUtGRzMwSm1QREo3ZzVfZlhJdDlCVy13eHpIMG1ucWxhcENfeWJDdkVCRE56RXI3cWJtSEZ1eGpkcEQyT3BweENaWnhheTZkYWtHZWRoeHRqZFNJZExHamVybnpCekQ1MnzjeaCVUm-W5lfjbbWCT-3ynIv9BlRcECqcDomzx9mJQA==
Error decoding string: illegal base64 data at input byte 258
([]uint8) (len=192 cap=222) {
 00000000  31 36 37 39 35 38 36 32  32 37 7c 75 70 72 6b 54  |1679586227|uprkT|
 00000010  35 51 52 69 70 30 71 33  6c 57 53 73 76 2d 47 6e  |5QRip0q3lWSsv-Gn|
 00000020  31 61 4a 31 72 6a 56 6d  62 4f 55 59 31 7a 45 47  |1aJ1rjVmbOUY1zEG|
 00000030  54 6f 48 47 4a 55 5f 70  74 43 79 43 53 37 70 5a  |ToHGJU_ptCyCS7pZ|
 00000040  49 6c 45 56 4c 6d 69 72  68 66 64 72 4a 51 49 45  |IlEVLmirhfdrJQIE|
 00000050  6e 44 64 4f 75 4b 4a 35  6d 67 45 4b 46 47 33 30  |nDdOuKJ5mgEKFG30|
 00000060  4a 6d 50 44 4a 37 67 35  5f 66 58 49 74 39 42 57  |JmPDJ7g5_fXIt9BW|
 00000070  2d 77 78 7a 48 30 6d 6e  71 6c 61 70 43 5f 79 62  |-wxzH0mnqlapC_yb|
 00000080  43 76 45 42 44 4e 7a 45  72 37 71 62 6d 48 46 75  |CvEBDNzEr7qbmHFu|
 00000090  78 6a 64 70 44 32 4f 70  70 78 43 5a 5a 78 61 79  |xjdpD2OppxCZZxay|
 000000a0  36 64 61 6b 47 65 64 68  78 74 6a 64 53 49 64 4c  |6dakGedhxtjdSIdL|
 000000b0  47 6a 65 72 6e 7a 42 7a  44 35 32 7c e3 79 a0 95  |GjernzBzD52|.y..|
}
(string) (len=380) "challenge=MTY3OTU4NjIyN3x1cHJrVDVRUmlwMHEzbFdTc3YtR24xYUoxcmpWbWJPVVkxekVHVG9IR0pVX3B0Q3lDUzdwWklsRVZMbWlyaGZkckpRSUVuRGRPdUtKNW1nRUtGRzMwSm1QREo3ZzVfZlhJdDlCVy13eHpIMG1ucWxhcENfeWJDdkVCRE56RXI3cWJtSEZ1eGpkcEQyT3BweENaWnhheTZkYWtHZWRoeHRqZFNJZExHamVybnpCekQ1MnzjeaCVUm-W5lfjbbWCT-3ynIv9BlRcECqcDomzx9mJQA==; Path=/; Expires=Sat, 22 Apr 2023 15:43:47 GMT; Max-Age=2592000"
(string) (len=12) "SetChallenge"
([]uint8) (len=64 cap=512) {
 00000000  b8 ce 95 33 b3 e1 e5 dc  0e 70 17 5d 8f 43 f0 ed  |...3.....p.].C..|
 00000010  f0 28 c2 36 14 3b 65 74  9d c2 4e 9b fc 05 6e ef  |.(.6.;et..N...n.|
 00000020  c1 e0 55 26 57 af b9 98  82 b3 aa bd b4 2d 59 12  |..U&W........-Y.|
 00000030  3a 5e b4 94 03 ec 4c 64  a4 54 b4 f8 8f 2b 2b 01  |:^....Ld.T...++.|
}
```

Notice that the output values are very different and not client side readable? Does this make sense @qfrank or am I missing something?

Reply:
> We've always took the challenge from the response body, because the value in the cookie is encrypted with a random key.

oh make sense, my bad 🤦‍♂️, i forgot it's encrypted. thank you for your detailed explanation 

> we don't use Set-Cookie anywhere in the code

we used it indirectly, that's one part how session works